File: lib/models/recurrence.dart
import 'dart:convert';

enum RecurrenceType { daily, weekly, monthly, yearly, custom }

class Recurrence {
  final RecurrenceType type;
  // custom: dias da semana
  final List<String>? weekdays; // ["seg","ter","qua","qui","sex","sab","dom"]
  // custom: intervalo numérico
  final int? interval;
  final String? intervalUnit; // "dias" | "semanas" | "meses"

  const Recurrence({
    required this.type,
    this.weekdays,
    this.interval,
    this.intervalUnit,
  });

  // ── Serialization ──────────────────────────────────────────────────────────

  Map<String, dynamic> toJson() {
    final map = <String, dynamic>{'tipo': _typeStr};
    if (weekdays != null && weekdays!.isNotEmpty) map['dias'] = weekdays;
    if (interval != null) map['intervalo'] = interval;
    if (intervalUnit != null) map['unidade'] = intervalUnit;
    return map;
  }

  String toJsonString() => jsonEncode(toJson());

  static Recurrence? fromJsonString(String? raw) {
    if (raw == null || raw.isEmpty) return null;
    try {
      final map = jsonDecode(raw) as Map<String, dynamic>;
      final tipo = map['tipo'] as String?;
      if (tipo == null) return null;
      final type = _parseType(tipo);
      if (type == null) return null;
      return Recurrence(
        type: type,
        weekdays: (map['dias'] as List?)?.cast<String>(),
        interval: map['intervalo'] as int?,
        intervalUnit: map['unidade'] as String?,
      );
    } catch (_) {
      return null;
    }
  }

  // ── Display ────────────────────────────────────────────────────────────────

  String get displayLabel => switch (type) {
        RecurrenceType.daily => 'Todo dia',
        RecurrenceType.weekly => 'Toda semana',
        RecurrenceType.monthly => 'Todo mês',
        RecurrenceType.yearly => 'Todo ano',
        RecurrenceType.custom => _customLabel,
      };

  String get _customLabel {
    if (weekdays != null && weekdays!.isNotEmpty) {
      const labels = {
        'seg': 'Seg', 'ter': 'Ter', 'qua': 'Qua',
        'qui': 'Qui', 'sex': 'Sex', 'sab': 'Sáb', 'dom': 'Dom',
      };
      return weekdays!.map((d) => labels[d] ?? d).join(', ');
    }
    if (interval != null && intervalUnit != null) {
      return 'A cada $interval $intervalUnit';
    }
    return 'Personalizado';
  }

  // ── Next date calculation ──────────────────────────────────────────────────

  DateTime? nextDate(DateTime from) {
    switch (type) {
      case RecurrenceType.daily:
        return from.add(const Duration(days: 1));
      case RecurrenceType.weekly:
        return from.add(const Duration(days: 7));
      case RecurrenceType.monthly:
        return DateTime(from.year, from.month + 1, from.day);
      case RecurrenceType.yearly:
        return DateTime(from.year + 1, from.month, from.day);
      case RecurrenceType.custom:
        if (interval != null && intervalUnit != null) {
          return switch (intervalUnit) {
            'dias' => from.add(Duration(days: interval!)),
            'semanas' => from.add(Duration(days: interval! * 7)),
            'meses' => DateTime(from.year, from.month + interval!, from.day),
            _ => null,
          };
        }
        if (weekdays != null && weekdays!.isNotEmpty) {
          const dayMap = {
            'seg': 1, 'ter': 2, 'qua': 3,
            'qui': 4, 'sex': 5, 'sab': 6, 'dom': 7,
          };
          for (int i = 1; i <= 7; i++) {
            final next = from.add(Duration(days: i));
            if (weekdays!.any((d) => dayMap[d] == next.weekday)) return next;
          }
        }
        return null;
    }
  }

  // ── Helpers ────────────────────────────────────────────────────────────────

  String get _typeStr => switch (type) {
        RecurrenceType.daily => 'diario',
        RecurrenceType.weekly => 'semanal',
        RecurrenceType.monthly => 'mensal',
        RecurrenceType.yearly => 'anual',
        RecurrenceType.custom => 'personalizado',
      };

  static RecurrenceType? _parseType(String tipo) => switch (tipo) {
        'diario' => RecurrenceType.daily,
        'semanal' => RecurrenceType.weekly,
        'mensal' => RecurrenceType.monthly,
        'anual' => RecurrenceType.yearly,
        'personalizado' => RecurrenceType.custom,
        _ => null,
      };
}

File: stacked-web/lib/utils/recurrence.ts
/** Paridade lib/models/recurrence.dart */
export type RecurrenceType = "daily" | "weekly" | "monthly" | "yearly" | "custom"

export type Recurrence = {
  type: RecurrenceType
  weekdays?: string[]
  interval?: number
  intervalUnit?: "dias" | "semanas" | "meses"
}

const TYPE_TO_DB: Record<RecurrenceType, string> = {
  daily: "diario",
  weekly: "semanal",
  monthly: "mensal",
  yearly: "anual",
  custom: "personalizado",
}

const DB_TO_TYPE: Record<string, RecurrenceType> = {
  diario: "daily",
  semanal: "weekly",
  mensal: "monthly",
  anual: "yearly",
  personalizado: "custom",
}

const WEEKDAY_MAP: Record<string, number> = {
  seg: 1,
  ter: 2,
  qua: 3,
  qui: 4,
  sex: 5,
  sab: 6,
  dom: 7,
}

export function parseRecurrence(raw: unknown): Recurrence | null {
  if (raw == null) return null
  let map: Record<string, unknown>
  if (typeof raw === "string") {
    const trimmed = raw.trim()
    if (!trimmed) return null
    try {
      map = JSON.parse(trimmed) as Record<string, unknown>
    } catch {
      return null
    }
  } else if (typeof raw === "object") {
    map = raw as Record<string, unknown>
  } else {
    return null
  }

  const tipo = map.tipo
  if (typeof tipo !== "string") return null
  const type = DB_TO_TYPE[tipo]
  if (!type) return null

  return {
    type,
    weekdays: Array.isArray(map.dias) ? map.dias.map(String) : undefined,
    interval: typeof map.intervalo === "number" ? map.intervalo : undefined,
    intervalUnit:
      map.unidade === "dias" || map.unidade === "semanas" || map.unidade === "meses"
        ? map.unidade
        : undefined,
  }
}

export function recurrenceToJson(recurrence: Recurrence): string {
  const map: Record<string, unknown> = { tipo: TYPE_TO_DB[recurrence.type] }
  if (recurrence.weekdays?.length) map.dias = recurrence.weekdays
  if (recurrence.interval != null) map.intervalo = recurrence.interval
  if (recurrence.intervalUnit) map.unidade = recurrence.intervalUnit
  return JSON.stringify(map)
}

export function recurrenceLabel(recurrence: Recurrence): string {
  switch (recurrence.type) {
    case "daily":
      return "Todo dia"
    case "weekly":
      return "Toda semana"
    case "monthly":
      return "Todo mês"
    case "yearly":
      return "Todo ano"
    case "custom":
      if (recurrence.weekdays?.length) {
        const labels: Record<string, string> = {
          seg: "Seg",
          ter: "Ter",
          qua: "Qua",
          qui: "Qui",
          sex: "Sex",
          sab: "Sáb",
          dom: "Dom",
        }
        return recurrence.weekdays.map((day) => labels[day] ?? day).join(", ")
      }
      if (recurrence.interval != null && recurrence.intervalUnit) {
        return `A cada ${recurrence.interval} ${recurrence.intervalUnit}`
      }
      return "Personalizado"
  }
}

export function computeNextRecurrenceDate(from: Date, recurrence: Recurrence): Date | null {
  const base = new Date(from.getFullYear(), from.getMonth(), from.getDate())

  switch (recurrence.type) {
    case "daily":
      return addDays(base, 1)
    case "weekly":
      return addDays(base, 7)
    case "monthly":
      return new Date(base.getFullYear(), base.getMonth() + 1, base.getDate())
    case "yearly":
      return new Date(base.getFullYear() + 1, base.getMonth(), base.getDate())
    case "custom":
      if (recurrence.interval != null && recurrence.intervalUnit) {
        switch (recurrence.intervalUnit) {
          case "dias":
            return addDays(base, recurrence.interval)
          case "semanas":
            return addDays(base, recurrence.interval * 7)
          case "meses":
            return new Date(
              base.getFullYear(),
              base.getMonth() + recurrence.interval,
              base.getDate(),
            )
        }
      }
      if (recurrence.weekdays?.length) {
        for (let i = 1; i <= 7; i++) {
          const next = addDays(base, i)
          const weekday = next.getDay() === 0 ? 7 : next.getDay()
          if (recurrence.weekdays.some((day) => WEEKDAY_MAP[day] === weekday)) {
            return next
          }
        }
      }
      return null
  }
}

function addDays(date: Date, days: number): Date {
  const next = new Date(date)
  next.setDate(next.getDate() + days)
  return next
}
